File: 654.maximum-binary-tree.java
/*
 * @lc app=leetcode id=654 lang=java
 *
 * [654] Maximum Binary Tree
 */

// @lc code=start

import java.util.Arrays;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {

    public TreeNode constructMaximumBinaryTree(int[] nums) {
        if (nums.length == 1)
            return new TreeNode(nums[0]);
        int delim = 0;
        int max = nums[0];
        for (int ix = 1; ix != nums.length; ++ix) {
            if (nums[ix] > max) {
                max = nums[ix];
                delim = ix;
            }
        }
        TreeNode root = new TreeNode(nums[delim]);
        if (delim != 0) {
            int[] left = Arrays.copyOfRange(nums, 0, delim);
            root.left = constructMaximumBinaryTree(left);
        }
        if (delim + 1 != nums.length) {
            int[] right = Arrays.copyOfRange(nums, delim + 1, nums.length);
            root.right = constructMaximumBinaryTree(right);
        }
        return root;
    }
}
// @lc code=end

File: 654.maximum-binary-tree.cpp
/*
 * @lc app=leetcode id=654 lang=cpp
 *
 * [654] Maximum Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        return constructMaximumBinaryTree(nums.cbegin(), nums.cend());
    }

    TreeNode *constructMaximumBinaryTree(
        vector<int>::const_iterator begin, 
        vector<int>::const_iterator end) const {
        if (begin == end) return nullptr;
        const auto pos = max_element(begin, end);
        TreeNode *root = new TreeNode { *pos };
        root->left = constructMaximumBinaryTree(begin, pos);
        root->right = constructMaximumBinaryTree(pos + 1, end);
        return root;
    }
};
// @lc code=end
